### PR TITLE
WIP: Core protocols

### PR DESCRIPTION
**Pull request by dialelo**

### CODE REVIEW

**Comment by dialelo** [comment]
Ready for a first release, @niwinz please review and merge.

**Comment by niwinz** [comment]
Really awesome work! Thanks!

File: src/language-advanced.adoc [hunk 1]
```diff
@@ -1199,12 +1199,701 @@ TBD
...
+
+[source, clojure]
+----
+(extend-type PersistentQueue
+  IFn
+  (-invoke
+    ([this idx]
+      (nth this idx))))
+
+(def q #queue[0 1 2])
```
Comment (wodin) [inline src/language-advanced.adoc]: This example might be clearer if the keys did not equal their corresponding values.

Comment (dialelo) [inline src/language-advanced.adoc]: Good point, I'm going to change so it doesn't lead to confusion. Thanks @wodin!

**Pull request merged**